### ISSUE TITLE
v2: Expose TLS placeholders

### ISSUE THREAD

**Issue by moorereason** [caddyserver/caddy#2977]
## Proposal

Born out of https://github.com/caddyserver/caddy/issues/2505#issuecomment-574224265.

### TLS Placeholders

```text
{http.request.tls.version}
	The TLS version name (e.g. "tls1.3")
{http.request.tls.cipher_suite}
	The TLS cipher suite (e.g. "TLS_CHACHA20_POLY1305_SHA256").
{http.request.tls.resumed}
	The TLS connection resumed a previous connection (e.g. "true").
{http.request.tls.proto}
	The negotiated next protocol (e.g. "h2").
{http.request.tls.proto_mutual}
	The negotiated next protocol was advertised by server (e.g. "true").
{http.request.tls.server_name}
	The server name requested by client, if any.

{http.request.tls.client.cert}
	The client certificate in PEM format.
{http.request.tls.client.fingerprint}
	The SHA256 checksum of the client certificate.
{http.request.tls.client.issuer}
	The issuer DN string of the client certificate.
{http.request.tls.client.serial}
	The serial number of the client certificate.
{http.request.tls.client.subject}
	The subject DN string of the client certificate.
```

### Shorthand Placeholders

| Shorthand  | Replaces                        |
|------------|---------------------------------|
| {tls_cipher}      |  {http.request.tls.cipher_suite}              |
| {tls_version}      |  {http.request.tls.version}              |
| {tls_client_cert}      |  {http.request.tls.client.cert}              |
| {tls_client_fingerprint}      |  {http.request.tls.client.fingerprint}              |
| {tls_client_issuer}      |  {http.request.tls.client.issuer}              |
| {tls_client_serial}      |  {http.request.tls.client.serial}              |
| {tls_client_subject}      | {http.request.tls.client.subject}              |

### Comments

1. ~~It's not clear to me if I should be using dashes or underscores for these placeholders.~~
1. Should the "client" placeholders be renamed to "peer" to match `tls.ConnectionState`?
1. I've renamed several items to be clearer (`issuer` vs `i-dn`).

I'm willing to work on a PR once we agree on the details.

### Revisions 📓 

- Use underscores instead of hyphens.
- Changed "i.e." to "e.g."
- Add underscores instead of mashing words together
- Removed `http.request.tls.client.cert_escaped`
- Changed `http.request.tls.client.cert_raw` to `http.request.tls.client.cert`
- Removed `http.request.tls.client.valid_*`
- Note that the `http.request.tls.client.fingerprint` is a SHA256 checksum.
- Use lowercase for `http.request.tls.version` result.

### Removed Items 🗑 

```text
{http.request.tls.client.cert_escaped}
	The client certificate in PEM format (url-encoded).
{http.request.tls.client.valid_end}
	The end date of the client certificate (e.g. "Jan 02 15:04:05 2006 UTC").
{http.request.tls.client.valid_days}
	The number of days remaining before the end date of the client certificate.
{http.request.tls.client.valid_start}
	The start date of client certificate (e.g. "Jan 02 15:04:05 2006 UTC").
```

**Comment by mholt** [comment]
This is a great proposal, thanks!

> It's not clear to me if I should be using dashes or underscores for these placeholders.

Underscores. Definitely.

> Should the "client" placeholders be renamed to "peer" to match tls.ConnectionState?

Hmmmm maybe, let's revisit that later.

> I've renamed several items to be clearer (issuer vs i-dn).

Thank you 🙏 

A couple other suggestions:

- Just a nit, but in the docs, where "i.e." ("in other words") is used, I think "e.g." ("for example") would be better instead.
- Add underscores instead of mashing words together:
  - `ciphersuite` -> `cipher_suite`
  - `protomutual` -> `proto_mutual`
  - `servername` -> `server_name`
- Drop the escaped variants (URL encoding should be handled by the middleware using the replacer)
- I'm not sure about the utility of the `valid*` placeholders yet. Maybe drop them for now until we decide what is needed and what is useful. Most of the other fields are pretty easy/obvious to add, but dates/timestamps are tricky so I want to tackle those separately.

**Comment by moorereason** [comment]
I've updated the proposal based upon your comments, @mholt.  PTAL

> Just a nit...

You were right.  If it's worth doing, it's worth doing right.  👍

**Comment by mholt** [comment]
This is looking good, would be a simple and useful PR to create, and easy for me to review. Ship it!

**Comment by moorereason** [comment]
These replacements are definitely scoped to an individual HTTP request, so it makes sense as-is.  I'm really new to Caddy, so I don't know the larger context of how replacements are used.  Is there any scenario where you would reference TLS values that are not related to a HTTP request?

PS - I'm about to update my OP to note that the client.fingerprint value is a SHA256 checksum.

**Comment by moorereason** [comment]
Updated proposal to use lowercase result for `http.request.tls.version` in order to match `caddytls.SupportedProtocols`.

**Comment by moorereason** [comment]
Resolved in b0a491aec808652dfd65910ee192ab88c07ac99d
